src/client/csfle/state_machine.rs

```diff
@@ -346,3 +364,134 @@ fn raw_to_doc(raw: &RawDocument) -> Result<Document> {
     raw.try_into()
         .map_err(|e| Error::internal(format!("could not parse raw document: {}", e)))
 }
+
+#[cfg(feature = "azure-kms")]
+pub(crate) mod azure {
+    use bson::{rawdoc, RawDocumentBuf};
+    use serde::Deserialize;
+    use std::time::{Duration, Instant};
+    use tokio::sync::Mutex;
+
+    use crate::{
+        error::{Error, Result},
+        runtime::HttpClient,
+    };
+
+    #[derive(Debug)]
+    pub(crate) struct ExecutorState {
+        cached_access_token: Mutex<Option<CachedAccessToken>>,
+        http: HttpClient,
+        #[cfg(test)]
+        pub(crate) test_host: Option<(&'static str, u16)>,
+        #[cfg(test)]
+        pub(crate) test_param: Option<&'static str>,
+    }
+
+    impl ExecutorState {
+        pub(crate) fn new() -> Result<Self> {
+            const AZURE_IMDS_TIMEOUT: Duration = Duration::from_secs(10);
+            Ok(Self {
+                cached_access_token: Mutex::new(None),
+                http: HttpClient::with_timeout(AZURE_IMDS_TIMEOUT)?,
+                #[cfg(test)]
+                test_host: None,
+                #[cfg(test)]
+                test_param: None,
+            })
+        }
+
+        pub(crate) async fn get_token(&self) -> Result<RawDocumentBuf> {
+            let mut cached_token = self.cached_access_token.lock().await;
+            if let Some(cached) = &*cached_token {
+                if cached.expire_time.saturating_duration_since(Instant::now())
+                    > Duration::from_secs(60)
+                {
+                    return Ok(cached.token_doc.clone());
+                }
+            }
+            let token = self.fetch_new_token().await?;
+            let out = token.token_doc.clone();
+            *cached_token = Some(token);
+            Ok(out)
+        }
+
+        async fn fetch_new_token(&self) -> Result<CachedAccessToken> {
+            let now = Instant::now();
+            let server_response: ServerResponse = self
+                .http
+                .get_and_deserialize_json(self.make_url()?, &self.make_headers())
+                .await
+                .map_err(|e| Error::authentication_error("azure imds", &format!("{}", e)))?;
+            let expires_in_secs: u64 = server_response.expires_in.parse().map_err(|e| {
+                Error::authentication_error(
+                    "azure imds",
+                    &format!("invalid `expires_in` response field: {}", e),
+                )
+            })?;
+            #[allow(clippy::redundant_clone)]
```

Comment:
why is a redundant clone needed here?

Reply:
The `#[cfg(test)]` line below needs `server_response`, but without the `clone` this line is a partial move.